### PR TITLE
Lanczos Iteration and Stochastic Lanczos Trace Estimator

### PR DESCRIPTION
**Pull request by akij17**
Implementation of Stochastic Lanczos Quadrature (SLQ) method for Trace Estimation of implicitly available matrix. 
This update also includes an implementation of Lanczos iteration method to find m most useful eigenvectors and eigenvalues of Hermitian matrix. Lanczos iteration returns V (an orthonormal matrix) and T (a tridiagonal real symmetric matrix).

### CODE REVIEW

File: src/lanczos.jl [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    v::TV
+    vₒ::TV
+    ω::TV
+    α::T
+    β::T
+    m::I
+end
+
+struct LanczosResult
+    V::AbstractMatrix
```
Comment (mohdibntarek) [inline src/lanczos.jl:21]: Make type parameter.

File: src/lanczos.jl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+Create a Lanczos Workspace struct to use with Lanczos iteration.
+# Arguments
+    - `A` : Symmetric Positive Definite Matrix
+    - `v` : An arbitary vector for a function: f -> A*v
+    - `m` : Number of iterations
+"""
+function LanczosWorkspace(A, v, m)
+    n = size(v, 1)
+    m = m > n ? n : m
```
Comment (mohdibntarek) [inline src/lanczos.jl:35]: `min(m, n)`

File: src/lanczos.jl [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    V = zeros(eltype(A), n, m)
```
Comment (mohdibntarek) [inline src/lanczos.jl:36]: `Tnum = eltype(A)` and reuse `Tnum`.

File: src/lanczos.jl [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+Takes a LanczosWorkspace object and returns LanczosResult object.
+Access SymTridiagonal T by R.T and orthonormal matrix V by R.V
+
+Example:
+lw = LanczosWorkspace(A, v, m)
+R = lanczos(lw)
+"""
+function lanczos(lw::LanczosWorkspace)
+    for i in 1:lw.m
```
Comment (mohdibntarek) [inline src/lanczos.jl:56]: Use Parameters.jl to `@unpack` and `@pack` all the fields of the workspace before and after the loop respectively.

File: src/lanczos.jl [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+# Arguments
+    - `A` : Symmetric Positive Definite Matrix
+    - `v` : An arbitary vector for a function: f -> A*v
+    - `m` : Number of iterations
+"""
+function LanczosWorkspace(A, v, m)
+    n = size(v, 1)
+    m = m > n ? n : m
+    V = zeros(eltype(A), n, m)
+    T = zeros(eltype(A), m, m)
```
Comment (mohdibntarek) [inline src/lanczos.jl:37]: This will be symmetric tridiagonal. No need to allocate a dense matrix. Check `SymTridiagonal`.

File: src/lanczos.jl [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        lw.α = dot(lw.ω, lw.v)
+        lw.ω = lw.ω - (lw.α * lw.v) - (lw.β * lw.vₒ)
+        lw.β = norm(lw.ω)
+        lw.T[i, i] = lw.α
+        if i < lw.m
+            lw.T[i, i+1] = lw.β
+            lw.T[i+1, i] = lw.β
+        end
+        lw.vₒ= lw.v
+        lw.v = lw.ω/lw.β
```
Comment (mohdibntarek) [inline src/lanczos.jl:67]: Use inplace version `lw.v .= lw.ω ./ lw.β`.

File: src/lanczos.jl [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+Example:
+lw = LanczosWorkspace(A, v, m)
+R = lanczos(lw)
+"""
+function lanczos(lw::LanczosWorkspace)
+    for i in 1:lw.m
+        mul!(lw.ω, lw.A, lw.v)
+        lw.α = dot(lw.ω, lw.v)
+        lw.ω = lw.ω - (lw.α * lw.v) - (lw.β * lw.vₒ)
```
Comment (mohdibntarek) [inline src/lanczos.jl:59]: Use inplace version with dots which will fuse all the operations in a single loop.

File: src/lanczos.jl [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        lw.β = norm(lw.ω)
+        lw.T[i, i] = lw.α
+        if i < lw.m
+            lw.T[i, i+1] = lw.β
+            lw.T[i+1, i] = lw.β
+        end
+        lw.vₒ= lw.v
```
Comment (mohdibntarek) [inline src/lanczos.jl:66]: This means that `lw.vₒ` and `lw.v` will be pointing to the same vector. I think what you want is `lw.vₒ .= lw.v` or `copy!(lw.vₒ, lw.v)` so they can be 2 different vectors but with the same content.

File: src/slq.jl [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+ - `randfunc` : Function to generate arbitary random vectors for Lanczos iteration
+                (If none supplied, f(n) = Base.rand(-1:2:1, n))
+ - `ctol` : Convergence Tolerance
+ - `m` : Lanczos iterations
+ - `nv` : SLQ Iterations (if convergence is achieved SLQ stops before nv)
+"""
+function SLQWorkspace(A; f::Function=invfun, ctol=0.1, m = 50, nv = 30)
+    o = one(eltype(A) <: Integer ? Float64 : eltype(A))
+    randfunc(n) = Base.rand(-o:2*o:o, n)
+    return SLQWorkspace(A, f, randfunc, m, nv, ctol, zero(eltype(A)))
```
Comment (mohdibntarek) [inline src/slq.jl:40]: Should allocate the Lanczos workspace here and have it as a field. The whole point of workspaces is to reuse the memory which is allocated only once at the beginning.

File: src/slq.jl [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+w = SLQWorkspace(A, m = 20, nv = 10)
+trace = slq(w)
+"""
+function slq(w::SLQWorkspace; skipverify = false)
+    if skipverify || isposdef(w.A)
+        tr = zero(eltype(w.A))
+        for i in 1:w.nᵥ
+            v₀ = w.randfunc(size(w.A, 1))
+            v₀ = v₀/norm(v₀)
+            lw = LanczosWorkspace(w.A, v₀, w.m)
```
Comment (mohdibntarek) [inline src/slq.jl:65]: Can make a function to update the Lanczos workspace inplace and use it here.

File: src/lanczos.jl [hunk 11]
```diff
@@ -0,0 +1,71 @@
+# Lαnczos iterαtion
+# Based on Lanczos method by C. Lanczos
+
+using LinearAlgebra
+
+export LanczosWorkspace, lanczos
+
+mutable struct LanczosWorkspace{T, TM <:AbstractMatrix{T}, TN <: AbstractMatrix{T}, TV <:AbstractVector{T}, I <:Integer}
+    A::TM
+    V::TN
```
Comment (mohdibntarek) [inline src/lanczos.jl:10]: Do we need to store the whole `V`? We only need the first row, no?

**Review by mohdibntarek — COMMENTED** [review]
Looks good. I left some comments to consider.

File: src/slq.jl [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+            R = lanczos(lw)
+            Tvec = eigvecs(R.T)
+            Tval = eigvals(R.T)
```
Comment (mohdibntarek) [inline src/slq.jl:68]: Make sure `R.T` is `SymTridiagonal`, and that this dispatches to an efficient method.

Comment (mohdibntarek) [reply src/slq.jl:68]: Should scale linearly with the size of `T`.

**Comment by akij17** [comment]
Late Update: Really busy with final term examinations before the summer vacations. Although most fixes are done there are still some problems relating to reducing the loops into a single operation. Will get back to it on May 5 (when the exams end).

**Pull request closed without merging**